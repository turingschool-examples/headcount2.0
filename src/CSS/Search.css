.search-bar {
  margin: 1rem 0.5rem;
}

.search-input {
  height: 2rem;
  width: 20rem;
  border: 1.5px solid #ccc;
  margin: 1rem 0;
  border-radius: 5px;
}

.search-input:hover {
  box-shadow: 1px 1px 1px 1px #ccc;
}

.hide-suggestions {
  display: none;
}

.did-you-mean {
  font-style: italic;
  color: gray;
}

.show-suggestions {
  margin: 1rem 0;
  text-align: center;
}

.suggestions {
  cursor: pointer;
  border: 1px solid gray;
  margin: 0;
  padding: 2px;
  text-align: center;
}

.suggestions:hover {
  background-color: rgb(244, 243, 243);
  box-shadow: 1.5px 1.5px 1.5px 1.5px #ccc;
}

button {
  width: 21.3rem;
  height: 2rem;
  border-radius: 5px;
  background-color: rgb(230, 230, 230);
  font-size: 1.1rem;
}

button:hover {
  background-color: rgb(142, 142, 142);
  color: rgb(237, 237, 237);
  cursor: pointer;
}

input[type="text"] {
  font-size: 1.1rem;
  line-height: 5rem;
  padding-left: 1rem;
}

.show {
  margin-bottom: 10rem;
  font-style: italic;
}

.show-none {
  display: none;
}

.display-error {
  padding: 1rem 0;
  text-align: center;
}
