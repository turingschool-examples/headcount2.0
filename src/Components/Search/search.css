.search-input{
  color: grey;
  margin: 0 auto;
  display: flex;
  width: 85%;
  height: 50px;
  font-size: 2rem;
  text-align: center;
  border-radius: 5px;
  border: 1px solid grey;
  box-shadow: 2px 2px 2px rgba(0, 0, 0, 0.5);
}

.submit-form{
  width: 85%;
  border-radius: 5px;
  display: flex;
  justify-content: center;
}

.search-container{
  width: 85%;
  display: flex;
  margin: 0 auto;
  justify-content: center;
}

.fixed-search{
  position: fixed;
}

.submit-button{
  width: 8em;
  height: 5em;
  border-radius: 5px;
  box-shadow: 2px 2px 2px rgba(0, 0, 0, 0.5);
}
