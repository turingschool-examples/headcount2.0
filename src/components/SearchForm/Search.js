import React, { Component } from 'react';
import PropTypes from 'prop-types';

export class Search extends Component {
  constructor() {
    super();
    this.state = {
      query: ''
    };
  }

  handleChange =  event => {
    const { value } = event.target;

    this.props.filterDistricts(value);

    this.setState({ query: value });
  }

  render() {
    const { query } = this.state;

    return (
      <input
        onChange={this.handleChange}
        type="text"
        name="query"
        value={query}
        placeholder="Search"
      />
    );
  }
}

Search.propTypes = {
  filterDistricts: PropTypes.func.isRequired
};

export default Search;
