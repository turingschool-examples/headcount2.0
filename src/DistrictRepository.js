import React, { Component } from 'react'
import SchoolList from './SchoolList'
import Search from './Search'
import kinderData from '../data/kindergartners_in_full_day_program'

export default class DistrictRepository extends Component {
  constructor(data) {
    super();
    this.data = this.getData(kinderData);
    this.state = {
      data: {}
    }
    this.findAllMatches = this.findAllMatches.bind(this);
  }

  componentDidMount() {
    this.setState({
      data: this.data
    })
  }

  getData(data) {
    return data.reduce((accu, elem) => {
      const newObj = Object.assign({}, elem, { Location: elem.Location.toUpperCase(), Data: {} } );

      if (!accu[elem.Location]) {
        accu[elem.Location] = newObj
      }

      const accuData = accu[elem.Location].Data;
      const roundedData = Math.round(1000*elem.Data)/1000;

      Object.assign(accuData, {[elem.TimeFrame]: roundedData || 0})

      return accu
    }, {})
  }

  findByName(name) {
    if (name === undefined) {
      return undefined
    }

    const upName = name.toUpperCase();
    const dataKey = Object.keys(this.data)
    const searchedKey = dataKey.filter( (key) => upName === key.toUpperCase() )
    return this.data[searchedKey]
  }

  findAllMatches(name) {
    const { data } = this
    const matchedArray = []
    const keys = Object.keys(data)
    keys.forEach( key => matchedArray.push(data[key]))

    if (!name) {
    return matchedArray
    }

    const newArray = matchedArray.filter( place => place.Location.toUpperCase().includes(name.toUpperCase()))

    this.setState({
      data: newArray
    })
    return newArray;
  }

  render() {
    const {data} = this.state
    return (
      <div>
        <Search findSchool={this.findAllMatches}/>
        <SchoolList data={data}/>
      </div>
    )
  }
}
