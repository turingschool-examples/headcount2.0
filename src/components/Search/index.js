import React, { Component } from 'react'
import PropTypes from 'prop-types'
import './Search.css'

class Search extends Component {
  constructor({findAllMatches}) {
    super()
    this.state = {
      query: ''
    }
  }

  handleInputChange = event => {
    const { value } = event.target

    this.setState({query: value})
    this.props.findAllMatches(value);
  }

  handleSubmit = event => {
    event.preventDefault()
  }

  render() {
    return(
      <form className="Search" onSubmit={this.handleSubmit} >
        <input
          className="search-input"
          aria-label="Search for a specific district here"
          placeholder="Search for a specific district"
          name="query"
          value={this.state.query}
          onChange={this.handleInputChange}
        />
      </form>
    )
  }
}

Search.propTypes = {
  findAllMatches: PropTypes.func.isRequired
}

export default Search