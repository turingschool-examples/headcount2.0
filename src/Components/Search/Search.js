import React, { Component } from 'react';
import $ from 'jquery';
import './search.css'
import PropTypes from 'prop-types';

class Search extends Component{
  constructor(){
    super();

    this.state={
      district: ''
    }
  }

  handleSubmit = (event) => {
    event.preventDefault();

    this.props.searchData(this.state.district);
  }

  handleInput = (event) => {
    const { name, value } = event.target;

    this.setState({
      [name]: value
    })
  }

  render(){
    return(
      <div className="search-container" onChange={(event)=>{this.handleSubmit(event)}}>
      <form className="submit-form">
      <input
      name="district"
      className="search-input"
      placeholder="Search Districts"
      onChange={this.handleInput} />
      <button className="submit-button">Search</button>
      </form>
      </div>
    )
  }
}

Search.propTypes = {
  searchData: PropTypes.func
}

export default Search;
