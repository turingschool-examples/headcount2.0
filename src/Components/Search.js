import React, { Component } from "react";
import DistrictRepository from "../helper";
import kindergarners from "../data/kindergartners_in_full_day_program";
import PropTypes from "prop-types";
import "../CSS/Search.css";

export default class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      districtInputOne: "",
      searchSuggestions: []
    };
    this.districtRepository = new DistrictRepository(kindergarners);
  }

  handleChange = ({ value }) => {
    this.setState({ districtInputOne: value });
    this.suggestDistricts();
  };

  suggestDistricts = () => {
    const { districtInputOne } = this.state;
    const searchSuggestions = this.districtRepository.findAllMatches(
      districtInputOne
    );
    this.props.selectedDistricts.forEach(district => {
      searchSuggestions.push(district);
    });

    this.setState({ searchSuggestions });
    if (districtInputOne.length > 1) {
      this.props.handleSubmit(searchSuggestions);
    } else {
      const newDistricts = this.districtRepository.districtsArray();
      this.props.handleSubmit(newDistricts);
    }
  };

  clearInput = () => {
    this.setState({ districtInputOne: "" });
    const newDistricts = this.districtRepository.districtsArray();
    this.props.handleSubmit(newDistricts);
  };

  searchDistrict = district => {
    const filteredDistricts = this.districtRepository.findByName(district);
    this.props.handleSubmit([filteredDistricts]);
  };

  render() {
    return (
      <div className="search-bar">
        <form>
          <input
            type="text"
            className="search-input"
            placeholder={
              this.props.selectedDistricts.length > 0
                ? "Remove Card to Search"
                : "Search District"
            }
            onChange={event => this.handleChange(event.target)}
            value={this.state.districtInputOne}
            disabled={this.props.selectedDistricts.length > 1}
          />
          <button onClick={() => this.props.clearComparisons()}>
            Clear Fields
          </button>
        </form>
        <div
          className={
            this.state.districtInputOne.length > 1
              ? "show-suggestions"
              : "hide-suggestions"
          }
        >
          <span
            className={
              this.state.searchSuggestions.length
                ? "show-none"
                : "display-error"
            }
          >
            None Found! Search Again
          </span>

          <span
            className={
              this.state.searchSuggestions.length ? "show" : "show-none"
            }
          >
            Did you Mean?
          </span>
          {this.state.searchSuggestions.map((district, index) => {
            if (
              this.state.searchSuggestions &&
              this.state.districtInputOne.length > 1
            ) {
              var suggestions = (
                <p
                  className="suggestions"
                  key={index}
                  onClick={event => {
                    this.props.selectCard(event.target.textContent);
                    this.clearInput();
                  }}
                >
                  {district.location}
                </p>
              );
            }
            return suggestions;
          })}
        </div>
      </div>
    );
  }
}

const { func, arrayOf, shape, bool, string, number } = PropTypes;

Search.propTypes = {
  selectedDistricts: arrayOf(
    shape({
      clicked: bool,
      location: string,
      stats: shape({ year: number })
    })
  ),
  handleSubmit: func.isRequired,
  selectCard: func.isRequired,
  clearComparisons: func.isRequired
};
