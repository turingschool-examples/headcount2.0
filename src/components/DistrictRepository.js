import React, { Component } from "react";
import Controls from "./Controls";
import { DistrictContainer } from "./DistrictContainer";
import KinderData from "../../data/kindergartners_in_full_day_program";
import "../styles/DistrictRepository.css";

export default class DistrictRepository extends Component {
  constructor(rawData) {
    super();
    // this.state.data = this.getDistrictData();
    this.state = {
      cards: [...this.findAllMatches()],
      data: this.getDistrictData()
    };
  }

  findAllMatches(userInput) {
    if (userInput) {
      userInput = userInput.toUpperCase();
    }

    const dataSet = this.getDistrictData();
    const districtKeys = Object.keys(dataSet).map(key => {
      return key.toUpperCase();
    });

    if (!userInput) {
      return districtKeys;
    }

    const filteredMatches = districtKeys.filter(district =>
      district.includes(userInput)
    );

    return filteredMatches;
  }

  findByName(userInput) {
    if (userInput) {
      userInput = userInput.toUpperCase();
    }
    const dataSet = this.state.data;
    const districtKeys = Object.keys(dataSet).map(key => {
      return key.toUpperCase();
    });

    let districtName = dataSet[userInput];
    let answer;

    if (districtKeys.includes(userInput)) {
      return (answer = {
        location: userInput,
        data: this.getData(districtName)
      });
    }
  }

  getData(districtInfo) {
    districtInfo = districtInfo.reduce((acc, item) => {
      if (item.Data === "N/A") {
        item.Data = 0;
      }
      acc[item.TimeFrame] = parseFloat(item.Data.toFixed(3));
      return acc;
    }, {});
    return districtInfo;
  }

  getDistrictData() {
    var rawData = KinderData;

    const cleanData = rawData.reduce((acc, obj) => {
      if (!acc[obj.Location]) {
        acc[obj.Location] = [];
      }
      if (acc[obj.Location]) {
        acc[obj.Location].push(obj);
      }
      return acc;
    }, {});
    return cleanData;
  }

  // componentDidMount()

  render() {
    return (
      <div className="district-repository-container">
        <Controls findDistrict={this.findByName.bind(this)} />
        <DistrictContainer
          foundData={this.state.cards}
          fullData={this.state.data}
        />
      </div>
    );
  }
}
