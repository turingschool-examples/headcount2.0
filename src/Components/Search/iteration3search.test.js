import React from 'react';
import { mount, shallow } from 'enzyme';
import ReactDOM from 'react-dom';
import Search from '../Search/Search';
import DistrictRepository from '../../helper.js';
import kinderData from '../../data/kindergartners_in_full_day_program.js';

describe('Search', () =>  {

  const mockData = {
      district: 'ADAMS'
    }

  const district = new DistrictRepository(kinderData);

  const handleInput = jest.fn();

  it('should match the snapshot', () => {

    const wrapper = shallow(<Search {...mockData} searchData={handleInput}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('should update state once input is handled', () => {

    const wrapper = shallow(<Search {...mockData} searchData={handleInput}/>);

    const mockEvent = {target: { value: 'ADAMS', name: 'district'} };

    const expectedState = { district: 'ADAMS' };

    wrapper.instance().handleInput(mockEvent);

    handleInput(mockEvent);

    expect(handleInput).toBeCalled();

    expect(wrapper.state()).toEqual(expectedState);

  });

  it('should update ', () => {

    const wrapper = shallow(<Search {...mockData} searchData={handleInput}/>);

    // wrapper.find('.search-container').simulate('submit', { preventDefault() {} });

  });

});
