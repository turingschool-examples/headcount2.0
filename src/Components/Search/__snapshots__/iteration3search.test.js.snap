// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search
    district="ADAMS"
    searchData={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        className="submit-form"
      >
        <input
          className="search-input"
          name="district"
          onChange={[Function]}
          placeholder="Search Districts"
        />
        <button
          className="submit-button"
        >
          Search
        </button>
      </form>,
      "className": "search-container",
      "onChange": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="search-input"
            name="district"
            onChange={[Function]}
            placeholder="Search Districts"
          />,
          <button
            className="submit-button"
          >
            Search
          </button>,
        ],
        "className": "submit-form",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "search-input",
            "name": "district",
            "onChange": [Function],
            "placeholder": "Search Districts",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search",
            "className": "submit-button",
          },
          "ref": null,
          "rendered": "Search",
          "type": "button",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          className="submit-form"
        >
          <input
            className="search-input"
            name="district"
            onChange={[Function]}
            placeholder="Search Districts"
          />
          <button
            className="submit-button"
          >
            Search
          </button>
        </form>,
        "className": "search-container",
        "onChange": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="search-input"
              name="district"
              onChange={[Function]}
              placeholder="Search Districts"
            />,
            <button
              className="submit-button"
            >
              Search
            </button>,
          ],
          "className": "submit-form",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "search-input",
              "name": "district",
              "onChange": [Function],
              "placeholder": "Search Districts",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search",
              "className": "submit-button",
            },
            "ref": null,
            "rendered": "Search",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
